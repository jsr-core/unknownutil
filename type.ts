/**
 * A type predicate function.
 *
 * ```ts
 * import { as, is, type Predicate } from "@core/unknownutil";
 *
 * type Person = {
 *   name: string;
 *   age: number;
 *   address?: string;
 * };
 * const isPerson = is.ObjectOf({
 *   name: is.String,
 *   age: is.Number,
 *   address: as.Optional(is.String),
 * }) satisfies Predicate<Person>;
 * ```
 */
export type Predicate<T> = (x: unknown) => x is T;

/**
 * A type predicated by {@linkcode Predicate<T>}.
 *
 * ```ts
 * import { as, is, type PredicateType } from "@core/unknownutil";
 *
 * const isPerson = is.ObjectOf({
 *   name: is.String,
 *   age: is.Number,
 *   address: as.Optional(is.String),
 * });
 *
 * type Person = PredicateType<typeof isPerson>;
 * // type Person = {
 * //   name: string;
 * //   age: number;
 * //   address?: string;
 * // };
 * ```
 */
export type PredicateType<P> = P extends Predicate<infer T> ? T : never;

/**
 * JavaScript primitive types.
 */
export type Primitive =
  | string
  | number
  | bigint
  | boolean
  | null
  | undefined
  | symbol;
