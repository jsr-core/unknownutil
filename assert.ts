import type { Predicate } from "./type.ts";

/**
 * A factory function that generates assertion error messages.
 */
export type AssertMessageFactory = (
  x: unknown,
  pred: Predicate<unknown>,
  name?: string,
) => string;

/**
 * The default factory function used to generate assertion error messages.
 */
export const defaultAssertMessageFactory: AssertMessageFactory = (
  x,
  pred,
  name,
) => {
  const p = pred.name || "anonymous predicate";
  const t = typeof x;
  const v = JSON.stringify(x, null, 2);
  return `Expected ${
    name ?? "a value"
  } that satisfies the predicate ${p}, got ${t}: ${v}`;
};

let assertMessageFactory = defaultAssertMessageFactory;

/**
 * Represents an error that occurs when an assertion fails.
 */
export class AssertError extends Error {
  /**
   * Constructs a new instance.
   * @param message The error message.
   */
  constructor(message?: string) {
    super(message);

    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, AssertError);
    }

    this.name = this.constructor.name;
  }
}

/**
 * Sets the factory function used to generate assertion error messages.
 *
 * ```ts
 * import { is, setAssertMessageFactory } from "@core/unknownutil";
 *
 * setAssertMessageFactory((x, pred) => {
 *   if (pred === is.String) {
 *     return `Expected a string, got ${typeof x}`;
 *   } else if (pred === is.Number) {
 *     return `Expected a number, got ${typeof x}`;
 *   } else if (pred === is.Boolean) {
 *     return `Expected a boolean, got ${typeof x}`;
 *   } else {
 *     return `Expected a value that satisfies the predicate, got ${typeof x}`;
 *   }
 * });
 * ```
 *
 * @param factory The factory function.
 */
export function setAssertMessageFactory(factory: AssertMessageFactory): void {
  assertMessageFactory = factory;
}

/**
 * Asserts that the given value satisfies the provided predicate.
 *
 * It throws {@linkcode AssertError} if the value does not satisfy the predicate.
 *
 * ```ts
 * import { assert, is } from "@core/unknownutil";
 *
 * const a: unknown = "hello";
 * assert(a, is.String);
 * // a is now narrowed to string
 * ```
 *
 * @param x The value to be asserted.
 * @param pred The predicate function to test the value against.
 * @param options Optional configuration for the assertion.
 * @returns The function has a return type of `asserts x is T` to help TypeScript narrow down the type of `x` after the assertion.
 */
export function assert<T>(
  x: unknown,
  pred: Predicate<T>,
  options: { message?: string; name?: string } = {},
): asserts x is T {
  if (!pred(x)) {
    throw new AssertError(
      options.message ?? assertMessageFactory(x, pred, options.name),
    );
  }
}
