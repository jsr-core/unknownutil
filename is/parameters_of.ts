import { rewriteName } from "../_funcutil.ts";
import type { WithOptional } from "../_annotation.ts";
import { hasOptional } from "../as/optional.ts";
import type { Predicate, PredicateType } from "../type.ts";
import { isArray } from "./array.ts";

/**
 * Return a type predicate function that returns `true` if the type of `x` is `ParametersOf<T>` or `ParametersOf<T, E>`.
 *
 * This is similar to {@linkcode isTupleOf}, but if {@linkcode asOptional} is specified at the trailing, the trailing elements becomes optional and makes variable-length tuple.
 *
 * To enhance performance, users are advised to cache the return value of this function and mitigate the creation cost.
 *
 * ```ts
 * import { as, is } from "@core/unknownutil";
 *
 * const isMyType = is.ParametersOf([
 *   is.Number,
 *   as.Optional(is.String),
 *   is.Boolean,
 *   as.Optional(is.Number),
 *   as.Optional(is.String),
 *   as.Optional(is.Boolean),
 * ] as const);
 * const a: unknown = [0, undefined, "a"];
 * if (isMyType(a)) {
 *   const _: [number, string | undefined, boolean, number?, string?, boolean?] = a;
 * }
 * ```
 *
 * With `predElse`:
 *
 * ```ts
 * import { as, is } from "@core/unknownutil";
 *
 * const isMyType = is.ParametersOf(
 *   [
 *     is.Number,
 *     as.Optional(is.String),
 *     as.Optional(is.Boolean),
 *   ] as const,
 *   is.ArrayOf(is.Number),
 * );
 * const a: unknown = [0, "a", true, 0, 1, 2];
 * if (isMyType(a)) {
 *   const _: [number, string?, boolean?, ...number[]] = a;
 * }
 * ```
 *
 * Depending on the version of TypeScript and how values are provided, it may be necessary to add `as const` to the array
 * used as `predTup`. If a type error occurs, try adding `as const` as follows:
 *
 * ```ts
 * import { as, is } from "@core/unknownutil";
 *
 * const predTup = [is.Number, is.String, as.Optional(is.Boolean)] as const;
 * const isMyType = is.ParametersOf(predTup);
 * const a: unknown = [0, "a"];
 * if (isMyType(a)) {
 *   const _: [number, string, boolean?] = a;
 * }
 * ```
 */
export function isParametersOf<
  T extends readonly [...Predicate<unknown>[]],
>(
  predTup: T,
): Predicate<ParametersOf<T>>;
export function isParametersOf<
  T extends readonly [...Predicate<unknown>[]],
  E extends Predicate<unknown[]>,
>(
  predTup: T,
  predElse: E,
): Predicate<[...ParametersOf<T>, ...PredicateType<E>]>;
export function isParametersOf<
  T extends readonly [...Predicate<unknown>[]],
  E extends Predicate<unknown[]>,
>(
  predTup: T,
  predElse?: E,
): Predicate<ParametersOf<T> | [...ParametersOf<T>, ...PredicateType<E>]> {
  const requiresLength = 1 +
    predTup.findLastIndex((pred) => !hasOptional(pred));
  if (!predElse) {
    return rewriteName(
      (x: unknown): x is ParametersOf<T> => {
        if (
          !isArray(x) || x.length < requiresLength || x.length > predTup.length
        ) {
          return false;
        }
        return predTup.every((pred, i) => pred(x[i]));
      },
      "isParametersOf",
      predTup,
    );
  } else {
    return rewriteName(
      (x: unknown): x is [...ParametersOf<T>, ...PredicateType<E>] => {
        if (!isArray(x) || x.length < requiresLength) {
          return false;
        }
        const head = x.slice(0, predTup.length);
        const tail = x.slice(predTup.length);
        return predTup.every((pred, i) => pred(head[i])) && predElse(tail);
      },
      "isParametersOf",
      predTup,
      predElse,
    );
  }
}

type ParametersOf<T> = T extends readonly [] ? []
  : T extends readonly [...infer P, infer R]
  // Tuple of predicates
    ? P extends Predicate<unknown>[]
      ? R extends Predicate<unknown> & WithOptional<unknown>
        // Last parameter is optional
        ? [...ParametersOf<P>, PredicateType<R>?]
        // Last parameter is NOT optional
      : [...ParametersOf<P>, PredicateType<R>]
    : never
  // Array of predicates
  : {
    -readonly [P in keyof T]: T[P] extends Predicate<infer U> ? U : never;
  };
