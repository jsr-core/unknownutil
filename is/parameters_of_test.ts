import { assertEquals } from "@std/assert";
import { assertSnapshot } from "@std/testing/snapshot";
import { assertType } from "@std/testing/types";
import type { Equal } from "../_testutil.ts";
import { as } from "../as/mod.ts";
import { is } from "./mod.ts";
import { isParametersOf } from "./parameters_of.ts";

Deno.test("isParametersOf<T>", async (t) => {
  await t.step("returns properly named predicate function", async (t) => {
    assertEquals(typeof isParametersOf([]), "function");
    await assertSnapshot(
      t,
      isParametersOf([is.Number, is.String, as.Optional(is.Boolean)]).name,
    );
    await assertSnapshot(
      t,
      isParametersOf([(_x): _x is string => false]).name,
    );
    await assertSnapshot(
      t,
      isParametersOf([]).name,
    );
    await assertSnapshot(
      t,
      isParametersOf([
        isParametersOf([
          isParametersOf([is.Number, is.String, as.Optional(is.Boolean)]),
        ]),
      ]).name,
    );
  });

  await t.step("returns true on T tuple", () => {
    const predTup = [is.Number, is.String, as.Optional(is.Boolean)] as const;
    assertEquals(isParametersOf(predTup)([0, "a", true]), true);
    assertEquals(isParametersOf(predTup)([0, "a"]), true);
    assertEquals(isParametersOf(predTup)([0, "a", undefined]), true);
  });

  await t.step("returns false on non T tuple", () => {
    const predTup = [is.Number, is.String, as.Optional(is.Boolean)] as const;
    assertEquals(isParametersOf(predTup)([0, 1, 2]), false);
    assertEquals(isParametersOf(predTup)([0]), false);
    assertEquals(isParametersOf(predTup)([0, "a", true, 0]), false);
  });

  await t.step("predicated type is correct", () => {
    const predTup = [
      as.Optional(is.Number),
      is.String,
      as.Optional(is.String),
      as.Optional(is.Boolean),
    ] as const;
    const a: unknown = [0, "a"];
    if (isParametersOf(predTup)(a)) {
      assertType<
        Equal<typeof a, [number | undefined, string, string?, boolean?]>
      >(true);
    }
  });
});

Deno.test("isParametersOf<T, R>", async (t) => {
  await t.step("returns properly named predicate function", async (t) => {
    assertEquals(typeof isParametersOf([], is.Array), "function");
    await assertSnapshot(
      t,
      isParametersOf([is.Number, is.String, as.Optional(is.Boolean)], is.Array)
        .name,
    );
    await assertSnapshot(
      t,
      isParametersOf([(_x): _x is string => false], is.ArrayOf(is.String))
        .name,
    );
    await assertSnapshot(
      t,
      isParametersOf([], is.ArrayOf(is.String)).name,
    );
    await assertSnapshot(
      t,
      isParametersOf([
        isParametersOf(
          [isParametersOf(
            [is.Number, is.String, as.Optional(is.Boolean)],
            is.Array,
          )],
          is.Array,
        ),
      ]).name,
    );
  });

  await t.step("returns true on T tuple", () => {
    const predTup = [is.Number, is.String, as.Optional(is.Boolean)] as const;
    const predRest = is.ArrayOf(is.Number);
    assertEquals(
      isParametersOf(predTup, predRest)([0, "a", true, 0, 1, 2]),
      true,
    );
    assertEquals(
      isParametersOf(predTup, predRest)([0, "a", undefined, 0, 1, 2]),
      true,
    );
    assertEquals(isParametersOf(predTup, predRest)([0, "a"]), true);
  });

  await t.step("returns false on non T tuple", () => {
    const predTup = [is.Number, is.String, as.Optional(is.Boolean)] as const;
    const predRest = is.ArrayOf(is.String);
    assertEquals(isParametersOf(predTup, predRest)("a"), false, "Not an array");
    assertEquals(
      isParametersOf(predTup, predRest)([0]),
      false,
      "Less than `predTup.length` - optional-count",
    );
    assertEquals(
      isParametersOf(predTup, predRest)([0, 1, 2]),
      false,
      "Not match `predTup` and no rest elements",
    );
    assertEquals(
      isParametersOf(predTup, predRest)([0, 1, 2, 0, 1, 2]),
      false,
      "Not match `predTup` and `predRest`",
    );
    assertEquals(
      isParametersOf(predTup, predRest)([0, "a", true, 0, 1, 2]),
      false,
      "Match `predTup` but not match `predRest`",
    );
    assertEquals(
      isParametersOf(predTup, predRest)([0, "a", undefined, 0, 1, 2]),
      false,
      "Match `predTup` but not match `predRest`",
    );
    assertEquals(
      isParametersOf(predTup, predRest)([0, "a", "b", "a", "b", "c"]),
      false,
      "Match `predRest` but not match `predTup`",
    );
    assertEquals(
      isParametersOf(predTup, predRest)([0, "a", "a", "b", "c"]),
      false,
      "Match `predRest` but no optional parameters",
    );
  });

  await t.step("predicated type is correct", () => {
    const predTup = [
      as.Optional(is.Number),
      is.String,
      as.Optional(is.String),
      as.Optional(is.Boolean),
    ] as const;
    const predRest = is.ArrayOf(is.Number);
    const a: unknown = [0, "a"];
    if (isParametersOf(predTup, predRest)(a)) {
      assertType<
        Equal<
          typeof a,
          [number | undefined, string, string?, boolean?, ...number[]]
        >
      >(
        true,
      );
    }
  });
});
