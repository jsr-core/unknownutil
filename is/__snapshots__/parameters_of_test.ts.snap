export const snapshot = {};

snapshot[`isParametersOf<T> > returns properly named predicate function 1`] = `
"isParametersOf([
  isNumber,
  isString,
  asOptional(isBoolean)
])"
`;

snapshot[`isParametersOf<T> > returns properly named predicate function 2`] = `"isParametersOf([(anonymous)])"`;

snapshot[`isParametersOf<T> > returns properly named predicate function 3`] = `"isParametersOf([])"`;

snapshot[`isParametersOf<T> > returns properly named predicate function 4`] = `
"isParametersOf([
  isParametersOf([
    isParametersOf([
      isNumber,
      isString,
      asOptional(isBoolean)
    ])
  ])
])"
`;

snapshot[`isParametersOf<T, R> > returns properly named predicate function 1`] = `
"isParametersOf([
  isNumber,
  isString,
  asOptional(isBoolean)
], isArray)"
`;

snapshot[`isParametersOf<T, R> > returns properly named predicate function 2`] = `"isParametersOf([(anonymous)], isArrayOf(isString))"`;

snapshot[`isParametersOf<T, R> > returns properly named predicate function 3`] = `"isParametersOf([], isArrayOf(isString))"`;

snapshot[`isParametersOf<T, R> > returns properly named predicate function 4`] = `
"isParametersOf([
  isParametersOf([
    isParametersOf([
      isNumber,
      isString,
      asOptional(isBoolean)
    ], isArray)
  ], isArray)
])"
`;
